RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc

```diff
@@ -0,0 +1,1030 @@
+		     << vertices.sw_[ivertex] << endl;
+	}
+      }
+#endif
+    }
+
+    auto osumpi = 1./sumpi;
+    for (unsigned int ivertex = 0; ivertex < nv; ++ivertex )
+      vertices.pk_[ ivertex ] = vertices.pk_[ ivertex ] * vertices.se_[ ivertex ] * osumpi;
+
+    return delta;
+  };
+  
+  delta += kernel_calc_zt(gvertices);
+  
+  // return how much the prototypes moved
+  return delta;
+}
+
+
+
+
+
+bool DAClusterizerInZT_vect::merge(vertex_t & y, double & beta)const{
+  // merge clusters that collapsed or never separated,
+  // only merge if the estimated critical temperature of the merged vertex is below the current temperature
+  // return true if vertices were merged, false otherwise
+  const unsigned int nv = y.getSize();
+
+  if (nv < 2)
+    return false;
+
+  // merge the smallest distance clusters first
+  std::vector<std::pair<double, unsigned int> > critical;
+  for (unsigned int k = 0; (k + 1) < nv; k++) {
+    if ( std::fabs(y.z_[k + 1] - y.z_[k]) < zmerge_ &&
+         std::fabs(y.t_[k + 1] - y.t_[k]) < tmerge_    ) {
+      auto dt2 = std::pow(y.t_[k + 1] - y.t_[k],2);
+      auto dz2 = std::pow(y.z_[k + 1] - y.z_[k],2);
+      critical.push_back( make_pair( dz2 + dt2, k ) );
+    }
+  }
+  if (critical.empty()) return false;
+
+  std::stable_sort(critical.begin(), critical.end(), std::less<std::pair<double, unsigned int> >() );
+
+
+  for (unsigned int ik=0; ik < critical.size(); ik++){
+    unsigned int k = critical[ik].second;
+    double rho = y.pk_[k]+y.pk_[k+1];
+    double swE = y.swE_[k]+y.swE_[k+1]-y.pk_[k]*y.pk_[k+1] / rho*( std::pow(y.z_[k+1]-y.z_[k],2) + std::pow(y.t_[k+1]-y.t_[k],2) );
+    double Tc = 2*swE / (y.sw_[k]+y.sw_[k+1]);
+
+    if(Tc*beta < 1){
+#ifdef VI_DEBUG
+      if(verbose_){ std::cout << "merging (" << y.z_[k + 1] << ',' << y.t_[k + 1] << ") and (" <<  y.z_[k] << ',' << y.t_[k] << ")  Tc = " << Tc <<  "  sw = "  << y.sw_[k]+y.sw_[k+1]  <<std::endl;}
+#endif
+      if(rho > 0){
+	y.z_[k] = (y.pk_[k]*y.z_[k] + y.pk_[k+1]*y.z_[k + 1])/rho;
+        y.t_[k] = (y.pk_[k]*y.t_[k] + y.pk_[k+1]*y.t_[k + 1])/rho;
+      }else{
+	y.z_[k] = 0.5 * (y.z_[k] + y.z_[k + 1]);
+        y.t_[k] = 0.5 * (y.t_[k] + y.t_[k + 1]);
+      }
+      y.pk_[k] = rho;
+      y.sw_[k] += y.sw_[k+1];
+      y.swE_[k] = swE;
+      y.removeItem(k+1);
+      return true;
+    }
+  }
+
+  return false;
+}
+
+
+
+
+bool 
+DAClusterizerInZT_vect::purge(vertex_t & y, track_t & tks, double & rho0, const double beta) const {
+  constexpr double eps = 1.e-100;
+  // eliminate clusters with only one significant/unique track
+  const unsigned int nv = y.getSize();
+  const unsigned int nt = tks.getSize();
+  
+  if (nv < 2)
+    return false;
+  
+  double sumpmin = nt;
+  unsigned int k0 = nv;
+  
+  int nUnique = 0;
+  double sump = 0;
+
+  std::vector<double> inverse_zsums(nt), arg_cache(nt), eik_cache(nt);
+  double * inverse_zsums_;
+  double * arg_cache_;
+  double * eik_cache_;
```

Comment:
please don't use the trailing underscores for local variables. This is easily confused with data members.